File: WeightedQUQF_part2/WeightedQuickUnionUF.py
class WeightedQuickUnionUF:
    def __init__(self, N):
        self.id = list(range(N))  # Each element points to itself
        self.size = [1] * N  # Initialize size array with 1

    def root(self, i):
        # First pass: Find the root
        while i != self.id[i]:
            i = self.id[i]
        root = i  # Store the root
        
        # Second pass: Path compression
        i = self.id[root]  # Reset to start compression
        while i != self.id[i]:
            next_id = self.id[i]  # Store next id
            self.id[i] = root     # Point directly to root
            i = next_id           # Move to next
        return root

    def connected(self, p, q):
        return self.root(p) == self.root(q)

    def union(self, p, q):
        i = self.root(p)
        j = self.root(q)
        if i == j:
            return
        # Weighted union: attach smaller tree under larger tree
        if self.size[i] < self.size[j]:
            self.id[i] = j
            self.size[j] += self.size[i]
        else:
            self.id[j] = i
            self.size[i] += self.size[j]

# Example usage
uf = WeightedQuickUnionUF(10)
uf.union(1, 2)
uf.union(3, 4)
uf.union(2, 3)

print("Connected(1, 3):", uf.connected(1, 3))  # True
print("Connected(1, 4):", uf.connected(1, 4))  # True
print("Connected(1, 5):", uf.connected(1, 5))  # False

File: WeightedQUQF_part2/WeightedQuickUnionUF.js
class WeightedQuickUnionUF {
    constructor(N) {
        this.id = Array.from({ length: N }, (_, i) => i); // Each element points to itself
        this.size = Array(N).fill(1); // Initialize size array with 1
    }

    root(i) {
        // First pass: Find the root
        while (i !== this.id[i]) {
            i = this.id[i];
        }
        const root = i; // Store the root
        
        // Second pass: Path compression
        i = this.id[root]; // Reset to start compression
        while (i !== this.id[i]) {
            const nextId = this.id[i]; // Store next id
            this.id[i] = root; // Point directly to root
            i = nextId; // Move to next
        }
        return root;
    }

    connected(p, q) {
        return this.root(p) === this.root(q);
    }

    union(p, q) {
        const i = this.root(p);
        const j = this.root(q);
        if (i === j) return;
        
        // Weighted union: attach smaller tree under larger tree
        if (this.size[i] < this.size[j]) {
            this.id[i] = j;
            this.size[j] += this.size[i];
        } else {
            this.id[j] = i;
            this.size[i] += this.size[j];
        }
    }
}

const uf = new WeightedQuickUnionUF(10);
uf.union(1, 2);
uf.union(3, 4);
uf.union(2, 3);

console.log("Connected(1, 3):", uf.connected(1, 3));  // true
console.log("Connected(1, 4):", uf.connected(1, 4));  // true
console.log("Connected(1, 5):", uf.connected(1, 5));  // false
